File: 2024/18/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	GRID_SIZE              = 71
	MAX_CORRUPTED_POSITIONS = 1024
)

var directions = [][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

func bfs(grid [][]rune, start, end [2]int) int {
	queue := [][2]int{start}
	visited := map[[2]int]bool{start: true}
	for steps := 0; len(queue) > 0; steps++ {
		nextQueue := [][2]int{}
		for _, pos := range queue {
			if pos == end {
				return steps
			}
			for _, d := range directions {
				next := [2]int{pos[0] + d[0], pos[1] + d[1]}
				if next[0] >= 0 && next[0] < GRID_SIZE && next[1] >= 0 && next[1] < GRID_SIZE && grid[next[0]][next[1]] == '.' && !visited[next] {
					visited[next] = true
					nextQueue = append(nextQueue, next)
				}
			}
		}
		queue = nextQueue
	}
	return -1
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var corruptedPositions [][2]int
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), ",")
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		corruptedPositions = append(corruptedPositions, [2]int{x, y})
	}

	grid := make([][]rune, GRID_SIZE)
	for i := range grid {
		grid[i] = make([]rune, GRID_SIZE)
		for j := range grid[i] {
			grid[i][j] = '.'
		}
	}
	for _, pos := range corruptedPositions[:MAX_CORRUPTED_POSITIONS] {
		grid[pos[0]][pos[1]] = '#'
	}

	fmt.Printf("Minimum steps to reach the exit: %d\n", bfs(grid, [2]int{0, 0}, [2]int{70, 70}))

	for _, pos := range corruptedPositions {
		grid[pos[0]][pos[1]] = '#'
		if bfs(grid, [2]int{0, 0}, [2]int{70, 70}) == -1 {
			fmt.Printf("First byte that blocks the path: (%d, %d)\n", pos[0], pos[1])
			break
		}
	}
}

File: 2024/18/main.js
const fs = require('fs');

const input = fs.readFileSync(__dirname + '/input.txt', 'utf8').trim().split('\n');
const corruptedPositions = input.map(line => line.split(',').map(Number));
const GRID_SIZE = 71, MAX_CORRUPTED_POSITIONS = 1024;
const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];

const bfs = (grid, start, end) => {
    const queue = [start], visited = new Set([`${start[0]},${start[1]}`]);
    for (let steps = 0; queue.length; steps++) {
        const nextQueue = [];
        for (const [x, y] of queue) {
            if (x === end[0] && y === end[1]) return steps;
            for (const [dx, dy] of directions) {
                const [nextX, nextY] = [x + dx, y + dy];
                if (nextX >= 0 && nextX < GRID_SIZE && nextY >= 0 && nextY < GRID_SIZE && grid[nextX][nextY] === '.' && !visited.has(`${nextX},${nextY}`)) {
                    visited.add(`${nextX},${nextY}`);
                    nextQueue.push([nextX, nextY]);
                }
            }
        }
        queue.splice(0, queue.length, ...nextQueue);
    }
    return -1;
};

const main = () => {
    const grid = Array.from({ length: GRID_SIZE }, () => Array(GRID_SIZE).fill('.'));
    corruptedPositions.slice(0, MAX_CORRUPTED_POSITIONS).forEach(([x, y]) => grid[x][y] = '#');
    
    console.log(`Minimum steps to reach the exit: ${bfs(grid, [0, 0], [70, 70])}`);

    for (const [x, y] of corruptedPositions) {
        grid[x][y] = '#';
        if (bfs(grid, [0, 0], [70, 70]) === -1) {
            console.log(`First byte that blocks the path: (${x}, ${y})`);
            break;
        }
    }
};

main();
